### PR TITLE
Initial baseline load-generator -> 5 home server tests.

### PR DESCRIPTION
**Pull request by marc-casavant**
The following new files include all files required to run a short and longer (5 minutes) running tests running on a baseline load-generator -> 5 home server multi-server environment.  Files include FreeRADIUS server related config files, a docker-compose environment and testcase configurations.

These new tests require the freeradius-multi-server repo to be available on the same system running the tests.

### CODE REVIEW

File: src/tests/multi-server/environments/configs/freeradius/homeserver/trigger.conf [hunk 1]
```diff
@@ -0,0 +1,224 @@
+#  -*- text -*-
```
Comment (alandekok) [inline src/tests/multi-server/environments/configs/freeradius/homeserver/trigger.conf]: It's probably worth removing everything you don't need from this file (and others).  i.e. there isn't value in having all of the document and comments in the test framework.

Plus, if the comments are duplicated here, they can get out of sync with the main configuration files.  so it's best to just simplify these examples.

The only comments in the configuration files here should be explanations as to why _this particular configuration_ was chosen.

File: src/tests/multi-server/environments/configs/freeradius/load-generator/sites-available/load-generator-5hs [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+
+        #
+		#  For now, only 'step' transport is available.
+		#
+        # process by proto_load_step.c
+        #####################################################################
+        transport = step
+        step {
+
+            # - filename (required)
```
Comment (alandekok) [inline src/tests/multi-server/environments/configs/freeradius/load-generator/sites-available/load-generator-5hs]: similar here, delete the default comments.  add a comment explaining why _this_ configuration has the given value.

File: src/tests/multi-server/environments/configs/freeradius/load-generator/sites-available/load-generator-5hs [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+			radius-module-homeserver2
+            radius-module-homeserver3
+            radius-module-homeserver4
+            radius-module-homeserver5
+		}
+    }
+
+    send Access-Request {
+        stats
+        ok
```
Comment (alandekok) [inline src/tests/multi-server/environments/configs/freeradius/load-generator/sites-available/load-generator-5hs]: Why is the `ok` here in every section?  It shouldn't be needed.

If the server authenticates the user, then everything should just work.

File: src/tests/multi-server/environments/configs/freeradius/load-generator/sites-available/load-generator-5hs [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+    }
+
+    # Accounting sections currently based on "default" virtual server config
+	recv Accounting-Request {
+		if (!Event-Timestamp) {
+			Event-Timestamp := %time.request() - %{Acct-Delay-Time || 0}
+		}
+		acct_unique
```
Comment (alandekok) [inline src/tests/multi-server/environments/configs/freeradius/load-generator/sites-available/load-generator-5hs]: this is likely not needed.  It's main use is when writing to SQL, and we're not doing that.

File: src/tests/multi-server/environments/configs/freeradius/load-generator/sites-available/load-generator-5hs [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+		files_accounting
+	}
+
+#    recv Accounting-Request {
+#        # Optional: use unlang to modify packet before calling radius module to send
+#        # requests to the upstream RADIUS server(s).
+#
+#		if (!Event-Timestamp) {
+#			Event-Timestamp := %time.request() - %{Acct-Delay-Time || 0}
```
Comment (alandekok) [inline src/tests/multi-server/environments/configs/freeradius/load-generator/sites-available/load-generator-5hs]: this is only needed for crappy NAS equipment.  It's better to just do:

```
Event.Timestamp := %timerequest()
```

since the timestamp changes every second, and therefore can't be put into a static packet.

File: src/tests/multi-server/environments/configs/freeradius/homeserver/clients.conf [hunk 6]
```diff
@@ -0,0 +1,11 @@
+client upstream-proxy {
+    ipaddr = $ENV{TEST_SUBNET}
+    secret = testing123
+    require_message_authenticator = auto
+    imit_proxy_state = auto
+    limit {
```
Comment (alandekok) [inline src/tests/multi-server/environments/configs/freeradius/homeserver/clients.conf]: the "limit" section is for TCP, and isn't needed for UDP packets.  You can delete it.

File: src/tests/multi-server/environments/configs/freeradius/homeserver/clients.conf [hunk 7]
```diff
@@ -0,0 +1,11 @@
+client upstream-proxy {
+    ipaddr = $ENV{TEST_SUBNET}
+    secret = testing123
+    require_message_authenticator = auto
+    imit_proxy_state = auto
```
Comment (alandekok) [inline src/tests/multi-server/environments/configs/freeradius/homeserver/clients.conf]: `*l*imit_proxy_state`

File: src/tests/multi-server/environments/configs/freeradius/homeserver/mods-available/linelog [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+        format = "This is a log message for %{User-Name}"
+        destination = file
+
+        file {
+                filename = ${logdir}/linelog
+                permissions = 0600
+                escape_filenames = no
+        }
+
+        pool {
```
Comment (alandekok) [inline src/tests/multi-server/environments/configs/freeradius/homeserver/mods-available/linelog]: I don't think we need a `pool` configuration for files?  You might need to check the linelog source code.

If the default `linelog` module comments aren't clear, we can update them to explain this issue.

File: src/tests/multi-server/environments/configs/freeradius/homeserver/sites-available/hs-auto-accept [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+	namespace = radius
+
+	radius {
+		Access-Request {
+			session {
+			}
+		}
+	}
+
+	dictionary {
```
Comment (alandekok) [inline src/tests/multi-server/environments/configs/freeradius/homeserver/sites-available/hs-auto-accept]: If you're not using this, just delete it.

File: src/tests/multi-server/environments/configs/freeradius/homeserver/sites-available/hs-auto-accept [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+		-ldap
+	}
+
+	authenticate eap {
+		eap
+	}
+
+	send Access-Challenge {
+		attr_filter.access_challenge
+		handled
```
Comment (alandekok) [inline src/tests/multi-server/environments/configs/freeradius/homeserver/sites-available/hs-auto-accept]: You can delete this line.  It's only needed for esoteric corner cases where the server is creating a challenge/response thing itself via `unlang`

File: src/tests/multi-server/environments/configs/freeradius/homeserver/sites-available/hs-auto-accept [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+	}
+
+	send Access-Reject {
+		%linelog_test_framework("hs-access-reject-sent - ERROR - %Y-%m-%d-%H:%G:%e Homeserver %{Net.Dst.IP}:%{Net.Dst.Port} tx Access-Accept src=%{Net.Src.IP}:%{Net.Src.Port} dst=%{Net.Dst.IP}:%{Net.Dst.Port}")
+		-sql
+		attr_filter.access_reject
+		eap
+		remove_reply_message_if_eap
+		delay_reject
```
Comment (alandekok) [inline src/tests/multi-server/environments/configs/freeradius/homeserver/sites-available/hs-auto-accept]: delete this. We don't need to delay rejects in the CI tests.

File: src/tests/multi-server/environments/configs/freeradius/homeserver/sites-available/hs-auto-accept [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+	}
+
+	recv Accounting-Request {
+		acct_unique
```
Comment (alandekok) [inline src/tests/multi-server/environments/configs/freeradius/homeserver/sites-available/hs-auto-accept]: delete this, as per other comments

File: src/tests/multi-server/environments/configs/freeradius/homeserver/sites-available/hs-auto-accept [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+		files_accounting
+	}
+
+	accounting Start {
+		-sql
```
Comment (alandekok) [inline src/tests/multi-server/environments/configs/freeradius/homeserver/sites-available/hs-auto-accept]: we're not using SQL, so just delete all `-sql` things, and leave the sections empty.

File: src/tests/multi-server/environments/configs/freeradius/load-generator/mods-available/linelog [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+        format = "This is a log message for %{User-Name}"
+        destination = file
+
+        file {
+                filename = ${logdir}/linelog
+                permissions = 0600
+                escape_filenames = no
+        }
+
+        pool {
```
Comment (alandekok) [inline src/tests/multi-server/environments/configs/freeradius/load-generator/mods-available/linelog]: the `pool` configuration is likely not needed.

File: src/tests/multi-server/environments/configs/freeradius/load-generator/mods-available/radius-5hs [hunk 15]
```diff
@@ -0,0 +1,449 @@
...
+			connection_timeout = 3.0
+			reconnect_delay = 5
+		}
+		request {
+			per_connection_max = 255
+			per_connection_target = 255
+			free_delay = 10
+		}
+	}
+	Access-Request {
```
Comment (alandekok) [inline src/tests/multi-server/environments/configs/freeradius/load-generator/mods-available/radius-5hs]: This is a large block of text which is copied multiple times.

perhaps create a directory `common/`, and then put the large common blocks there.  the configurations can then do `$INCLUDE common/foo.conf`

I'm a big fan of "don't repeat yourself".  If the same thing shows up 2-3 times, find a way to re-use a common block.

File: src/tests/multi-server/environments/configs/freeradius/load-generator/mods-available/radius-5hs [hunk 16]
```diff
@@ -0,0 +1,449 @@
...
+
+radius radius-module-homeserver4 {
+	mode = client
+	transport = udp
+	type = Access-Request
+	require_message_authenticator = auto
+	response_window = 15
+	zombie_period = 10
+	revive_interval = 3600
+	status_check {
```
Comment (alandekok) [inline src/tests/multi-server/environments/configs/freeradius/load-generator/mods-available/radius-5hs]: the status check configs can likely go into a common block, too.  Perhaps even more of the `radius` configuration block, too.

File: src/tests/multi-server/environments/configs/freeradius/load-generator/radiusd.conf [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+thread pool {
+}
+
+$INCLUDE trigger.conf
+
+global {
+        $INCLUDE global.d/
+}
+
+migrate {
```
Comment (alandekok) [inline src/tests/multi-server/environments/configs/freeradius/load-generator/radiusd.conf]: Delete this section.  It's long out of date.

File: src/tests/multi-server/environments/configs/freeradius/homeserver/trigger.conf (hunk 1 above)
Comment (marc-casavant) [reply src/tests/multi-server/environments/configs/freeradius/homeserver/trigger.conf]: Good point. Would we want to keep empty sections in the config file as placeholders?

In ../homeserver/radiusd.conf for example, we do not need to configure "thread pool". We could however refer to the config section:

```
thread pool {

}
```

File: src/tests/multi-server/README.md [hunk 18]
```diff
@@ -0,0 +1,31 @@
+Multi-server testcases requires the availability of the "freeradius-multi-server" test framework repo on your system.
+
+freeradius-multi-server repo: https://github.com/InkbridgeNetworks/freeradius-multi-server
+
+Multi-server environment Docker compose files are based on the following FreeRADIUS V4 Docker image:
+
+https://github.com/marc-casavant/freeradius-server/tree/dev-docker-build-from-src
+```bash
+freeradius-server % docker build --no-cache -t freeradius-dev-ubuntu24 -f scripts/docker/dev/build/ubuntu24/Dockerfile .
+```
```
Comment (arr2036) [inline src/tests/multi-server/README.md:20]: Get the crossbuild images to work on MacOS.  Either get arm versions of libkqueue, or submit a CI task for the libkqueue repo to build the .deb files from x86 and arm.

File: src/tests/multi-server/README.md [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+
+freeradius-multi-server % source .venv/bin/activate
+
+(.venv) freeradius-multi-server % TEST_LOGGER_CONFIG=linelog_file DATA_PATH=$HOME/sandbox/freeradius-server/src/tests/multi-server/environments/configs LISTENER_DIR=$HOME/freeradius-listener-logs docker compose -p custom_test-env-loadgen-5hs -f $HOME/sandbox/freeradius-server/src/tests/multi-server/environments/docker-compose/env-loadgen-5hs.yml up
+```
+
+### Multi-server test run
+
+```bash
+DATA_PATH=$HOME/sandbox/freeradius-server/src/tests/multi-server/environments/configs make test-framework-custom-config-path -- -x -vvvv --compose $HOME/sandbox/freeradius-server/src/tests/multi-server/environments/docker-compose/env-loadgen-5hs.yml --test $HOME/sandbox/freeradius-server/src/tests/multi-server/test-5hs-autoaccept.yml --use-files --listener-dir $HOME/freeradius-listener-logs
```
Comment (arr2036) [inline src/tests/multi-server/README.md]: Should be integrated into the FreeRADIUS make framework.  Add an all.mk that adds a `tests.multi` target, with correct prereqs to build the docker images using crossbuild.

File: src/tests/multi-server/README.md [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+Run Docker container based environment without starting the test; no traffic generated.
+
+```bash
+% cd $HOME/sandbox/freeradius-multi-server
+
+freeradius-multi-server % make configure
+
+freeradius-multi-server % source .venv/bin/activate
+
+(.venv) freeradius-multi-server % TEST_LOGGER_CONFIG=linelog_file DATA_PATH=$HOME/sandbox/freeradius-server/src/tests/multi-server/environments/configs LISTENER_DIR=$HOME/freeradius-listener-logs docker compose -p custom_test-env-loadgen-5hs -f $HOME/sandbox/freeradius-server/src/tests/multi-server/environments/docker-compose/env-loadgen-5hs.yml up
```
Comment (arr2036) [inline src/tests/multi-server/README.md]: This be should be created automatically in `build/tests/multi` so we can reinitialise the state easily.

File: src/tests/multi-server/README.md (hunk 19 above)
Comment (arr2036) [reply src/tests/multi-server/README.md]: Get Ethan to publish the multi-server test packages to PyPi.

Comment (arr2036) [reply src/tests/multi-server/README.md]: Any use of $HOME should be substituted for paths in `build/tests/multi`.

File: src/tests/multi-server/environments/configs/freeradius/load-generator/mods-available/radius-5hs (hunk 16 above)
Comment (arr2036) [reply src/tests/multi-server/environments/configs/freeradius/load-generator/mods-available/radius-5hs]: Just use templates :)

File: src/tests/multi-server/environments/configs/freeradius/homeserver/radiusd.conf [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+radacctdir = ${logdir}/radacct
+
+name = freeradius
+
+confdir = ${raddbdir}
+modconfdir = ${confdir}/mods-config
+certdir = ${confdir}/certs
+cadir   = ${confdir}/certs
+run_dir = ${localstatedir}/run/${name}
+db_dir = ${localstatedir}/lib/${name}
```
Comment (arr2036) [inline src/tests/multi-server/environments/configs/freeradius/homeserver/radiusd.conf]: Almost all the paths can be removed

File: src/tests/multi-server/environments/configs/freeradius/homeserver/radiusd.conf [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+
+libdir = /usr/lib/freeradius
+
+pidfile = ${run_dir}/${name}.pid
```
Comment (arr2036) [inline src/tests/multi-server/environments/configs/freeradius/homeserver/radiusd.conf]: Don't need a PID file

File: src/tests/multi-server/environments/configs/freeradius/homeserver/radiusd.conf [hunk 23]
```diff
@@ -0,0 +1,93 @@
...
+        file = ${logdir}/radius.log
+
+        syslog_facility = daemon
+}
+
+log debug {
+        destination = null
+        colourise = no
+        timestamp = yes
+}
```
Comment (arr2036) [inline src/tests/multi-server/environments/configs/freeradius/homeserver/radiusd.conf]: Can this be removed

File: src/tests/multi-server/environments/configs/freeradius/homeserver/radiusd.conf [hunk 24]
```diff
@@ -0,0 +1,93 @@
...
+ENV {
+
+}
+
+templates {
+        $INCLUDE template.d/
+}
+
+security {
+        user = freerad
```
Comment (arr2036) [inline src/tests/multi-server/environments/configs/freeradius/homeserver/radiusd.conf]: No need to set a user or group here, it can just run as root.

File: src/tests/multi-server/environments/configs/freeradius/homeserver/radiusd.conf [hunk 25]
```diff
@@ -0,0 +1,93 @@
...
+        allow_core_dumps = yes
+
+        max_attributes = 200
+}
+
+$INCLUDE clients.conf
+
+thread pool {
+
+}
```
Comment (arr2036) [inline src/tests/multi-server/environments/configs/freeradius/homeserver/radiusd.conf]: Remove empty sections

File: src/tests/multi-server/environments/configs/freeradius/homeserver/radiusd.conf.j2 [hunk 26]
```diff
@@ -0,0 +1,93 @@
...
+
+thread pool {
+
+}
+
+$INCLUDE trigger.conf
+
+global {
+        $INCLUDE global.d/
+}
```
Comment (arr2036) [inline src/tests/multi-server/environments/configs/freeradius/homeserver/radiusd.conf.j2:34]: No need for globals

File: src/tests/multi-server/environments/configs/freeradius/homeserver/sites-available/hs-auto-accept [hunk 27]
```diff
@@ -0,0 +1,144 @@
...
+			session {
+			}
+		}
+	}
+
+	listen authentication {
+		type = Access-Request
+		type = Status-Server
+		transport = udp
+		require_message_authenticator = auto
```
Comment (arr2036) [inline src/tests/multi-server/environments/configs/freeradius/homeserver/sites-available/hs-auto-accept]: Rs which are defaults

File: src/tests/multi-server/environments/docker-compose/env-loadgen-5hs.yml [hunk 28]
```diff
@@ -0,0 +1,274 @@
...
+
+      # Ensure EAP/PEAP certs exist (generate only if missing)
+      if [ ! -f /etc/raddb/certs/rsa/server.pem ]; then
+        echo "[*] Generating FreeRADIUS certs..."
+        (cd /etc/raddb/certs && make)
+      fi
+
+      rm -f /etc/raddb/sites-enabled/default
+      ln -sf /etc/raddb/sites-available/hs-auto-accept /etc/raddb/sites-enabled/hs-auto-accept
+      ln -sf /etc/raddb/mods-available/linelog_test_framework /etc/raddb/mods-enabled/linelog_test_framework
```
Comment (arr2036) [inline src/tests/multi-server/environments/docker-compose/env-loadgen-5hs.yml]: This can all be removed

File: src/tests/multi-server/environments/docker-compose/env-loadgen-5hs.yml [hunk 29]
```diff
@@ -0,0 +1,274 @@
...
+    - ${LISTENER_DIR}/:/var/run/multi-server/
+    command: radiusd -X
+    restart: unless-stopped
+    entrypoint:
+    - bash
+    - -lc
+    - |
+      set -euo pipefail
+
+      apt-get update
```
Comment (arr2036) [inline src/tests/multi-server/environments/docker-compose/env-loadgen-5hs.yml]: Remove this

File: src/tests/multi-server/test-5hs-autoaccept-5min.yml [hunk 30]
```diff
@@ -0,0 +1,80 @@
...
+              tail -f /dev/null
+
+            detach: true
+    verify:
+      timeout: 360
+      trigger_mode: unordered
+      triggers:
+      - hs-access-request-received:
+          may_pattern:
+            reg_pattern: \- INFO \- (\d+)-(\d+)-(\d+)-(\d+):(\d+):(\d+) Homeserver (\d+\.\d+\.\d+\.\d+):(\d+) rx Access-Request src=
```
Comment (arr2036) [inline src/tests/multi-server/test-5hs-autoaccept-5min.yml:61]: Emit structured data and use something like JQ, this is pretty fragile.

Comment (arr2036) [reply src/tests/multi-server/test-5hs-autoaccept-5min.yml:61]: Any linelog data should be a JSON blob.

File: src/tests/multi-server/environments/configs/freeradius/homeserver/radiusd.conf (hunk 23 above)
Comment (marc-casavant) [reply src/tests/multi-server/environments/configs/freeradius/homeserver/radiusd.conf]: `log debug {}` specifically or from the `request {}` section ... up to `log debug {}`?

File: src/tests/multi-server/environments/docker-compose/env-loadgen-5hs.yml (hunk 28 above)
Comment (marc-casavant) [reply src/tests/multi-server/environments/docker-compose/env-loadgen-5hs.yml]: I removed the soft link to enable the radius module since it is not needed.  However, the soft link to enable the linelog_test_framework config is required since it is a custom linelog configuration pulled in from the test framework.

File: src/tests/multi-server/environments/configs/freeradius/load-generator/template.d/load-generator-templates [hunk 31]
```diff
@@ -0,0 +1,77 @@
...
+			connection_timeout = 3.0
+			reconnect_delay = 5
+		}
+		request {
+			per_connection_max = 255
+			per_connection_target = 255
+			free_delay = 10
+		}
+	}
+	Access-Request {
```
Comment (alandekok) [inline src/tests/multi-server/environments/configs/freeradius/load-generator/template.d/load-generator-templates]: I don't think you need these?  The module has defaults which can be used.  But since it's used as a template, it's likely OK to leave it in.

File: src/tests/multi-server/environments/configs/freeradius/load-generator/radiusd.conf [hunk 32]
```diff
@@ -1,79 +1,326 @@
...
 }
 
-ENV {
+security {
+    allow_core_dumps = yes
 }
 
-templates {
-        $INCLUDE template.d/
+snmp {
```
Comment (alandekok) [inline src/tests/multi-server/environments/configs/freeradius/load-generator/radiusd.conf]: If we're not sending SNMP traps, this section can be deleted.

File: src/tests/multi-server/environments/configs/freeradius/load-generator/radiusd.conf [hunk 33]
```diff
@@ -1,79 +1,326 @@
...
+            radius-module-homeserver4
+            radius-module-homeserver5
+		}
+    }
+
+    send Access-Request {
+        stats
+    }
+
+    recv Access-Accept {
```
Comment (alandekok) [inline src/tests/multi-server/environments/configs/freeradius/load-generator/radiusd.conf]: ? The server isn't receiving Access-Accept.  This can be deleted.

File: src/tests/multi-server/environments/configs/freeradius/load-generator/radiusd.conf [hunk 34]
```diff
@@ -1,79 +1,326 @@
...
+        redundant-load-balance {
+			radius-module-homeserver1
+			radius-module-homeserver2
+            radius-module-homeserver3
+            radius-module-homeserver4
+            radius-module-homeserver5
+		}
+    }
+
+    send Access-Request {
```
Comment (alandekok) [inline src/tests/multi-server/environments/configs/freeradius/load-generator/radiusd.conf]: ? The server isn't sending Access-Request.  This can be deleted.

File: src/tests/multi-server/environments/configs/freeradius/load-generator/radiusd.conf [hunk 35]
```diff
@@ -1,79 +1,326 @@
...
+	}
+
+	accounting Accounting-Off {
+	}
+
+	accounting Failed {
+	}
+
+	send Accounting-Response {
+		detail
```
Comment (alandekok) [inline src/tests/multi-server/environments/configs/freeradius/load-generator/radiusd.conf]: Are the tests looking at the detail files?  If not, this can just be deleted.

File: src/tests/multi-server/environments/configs/freeradius/load-generator/radiusd.conf [hunk 36]
```diff
@@ -1,79 +1,326 @@
...
+		attr_filter.accounting_response
```
Comment (alandekok) [inline src/tests/multi-server/environments/configs/freeradius/load-generator/radiusd.conf]: This isn't needed, and can be deleted.

The "filter accounting response" is there to ensure RFC compliance.  i.e. "strip out attributes which will make the client upset".

For the tests, we don't create bad contents, and the client doesn't care what's in the reply.

File: src/tests/multi-server/environments/docker-compose/env-loadgen-5hs.yml (hunk 28 above)
Comment (arr2036) [reply src/tests/multi-server/environments/docker-compose/env-loadgen-5hs.yml]: In general all these homeserver entries should be removed except one, and that should be added into the middle of a Jinja loop.  This lets you parameterise the environment too, so it's N backends instead of 5.

File: src/tests/multi-server/environments/docker-compose/env-loadgen-5hs.yml [hunk 37]
```diff
@@ -0,0 +1,269 @@
...
+      touch freeradius-load-generator.log
+
+      source /tmp/env-setup.sh
+      export TEST_PROJECT_NAME=${COMPOSE_PROJECT_NAME}
+
+      # Calculate TEST_CONF_NUM_MESSAGES dynamically based on proto_load default env variables.
+      # Helps setup the Docker Compose environment in such a way that configures the load-generator
+      # to send a small amount of traffice for debugging purposes. All load-generator env variables
+      # can be overridden via the test config file.
+      export TEST_CONF_START_PPS=5
```
Comment (arr2036) [inline src/tests/multi-server/environments/docker-compose/env-loadgen-5hs.yml]: I don't get why these are here.  The environment should already be automatically re-exported to anything bash calls?

File: src/tests/multi-server/environments/docker-compose/env-loadgen-5hs.yml [hunk 38]
```diff
@@ -0,0 +1,269 @@
...
+      TEST_CONF_MAX_BACKLOG: 1000
+      TEST_CONF_REPEAT: "no"
+    entrypoint:
+    - bash
+    - -lc
+    - |
+      set -euo pipefail
+
+      # FreeRADIUS log file for load-generator
+      touch freeradius-load-generator.log
```
Comment (arr2036) [inline src/tests/multi-server/environments/docker-compose/env-loadgen-5hs.yml]: Why is this needed?  Is something not able to automatically create the file?  If that's the case, the code should be corrected to pass `O_CREAT` or whatever it needs.

File: src/tests/multi-server/environments/docker-compose/env-5hs-autoaccept.yml [hunk 39]
```diff
@@ -0,0 +1,269 @@
...
+      # Disable default virtual server config and disable unused modules.
+      # load-generator radiusd.conf has all necessary configurations.
+      rm -f /etc/raddb/sites-enabled/default
+      rm -f /etc/raddb/sites-enabled/inner-tunnel
+      rm -f /etc/raddb/mods-enabled/eap
+      rm -f /etc/raddb/mods-enabled/eap_inner
+      rm -f /etc/raddb/mods-enabled/mschap
+      rm -f /etc/raddb/mods-enabled/linelog
+
+      sleep infinity
```
Comment (arr2036) [inline src/tests/multi-server/environments/docker-compose/env-5hs-autoaccept.yml]: This is so radiusd can be invoked in a synchronised way from outside of the container?

File: src/tests/multi-server/environments/configs/freeradius/load-generator/template.d/load-generator-templates (hunk 31 above)
Comment (marc-casavant) [reply src/tests/multi-server/environments/configs/freeradius/load-generator/template.d/load-generator-templates]: Good point. I'll have a closer look and clean up the additional configs not needed for the load-generator.

File: src/tests/multi-server/environments/docker-compose/env-loadgen-5hs.yml (hunk 38 above)
Comment (marc-casavant) [reply src/tests/multi-server/environments/docker-compose/env-loadgen-5hs.yml]: It is to create a file in case we run ``freeradius -X` > freeradius-load-generator.log 2>&1 &` in the load-generator container.  By default, the test starts the load-generator using `freeradius -X` after the environment is ready but having the option of outputting the server logs is useful when debugging tests. 

If I don't run `touch freeradius-load-generator.log` at the entrypoint of the load-generator a `freeradius-load-generator.log` directory is created locally instead of the log file which causes runtime issues ... it's only used when debugging the server.

This is because one of the container volumes is:
`- ${DATA_PATH}/freeradius/load-generator/logs/freeradius-load-generator.log:/freeradius-load-generator.log`

File: src/tests/multi-server/environments/docker-compose/env-loadgen-5hs.yml (hunk 37 above)
Comment (marc-casavant) [reply src/tests/multi-server/environments/docker-compose/env-loadgen-5hs.yml]: These are only needed when debugging the load-generator while starting the server manually after starting the test environment.  We can certainly remove this but useful to have in certain troubleshooting scenarios.

File: src/tests/multi-server/environments/docker-compose/env-5hs-autoaccept.yml (hunk 39 above)
Comment (marc-casavant) [reply src/tests/multi-server/environments/docker-compose/env-5hs-autoaccept.yml]: Yes, this is the intent.  For proper functionality of the tests running the load-generator, it is important for the test to start the server after the environment is ready.  The `test-5hs-autoaccept.yml` testcase config is setup to have "state_1" configure the load-generator, wait for 50 seconds to allow the environment to finish its configuration/setup and then starts the server.

File: src/tests/multi-server/environments/docker-compose/env-loadgen-5hs.yml (hunk 38 above)
Comment (marc-casavant) [reply src/tests/multi-server/environments/docker-compose/env-loadgen-5hs.yml]: An option would be for to add an empty load-generator log file which would allow us to remove remove the "touch" command with this container volume:

`- ${DATA_PATH}/freeradius/load-generator/logs/freeradius-load-generator.log:/freeradius-load-generator.log`

I can certainly implement another solution, however, this optional log file is very useful while debugging the env.

**Comment by alandekok** [comment]
This has been merged.
